### PR TITLE
Use delta in direct write also, #22188

### PR DESCRIPTION
**Pull request by patriknw**
* Follow up on the causal delivery of deltas. #22331
* The first implementation used full state for the direct
  Write messages, i.e. updates with WriteConsistency != LocalWrite
* This is an optimization so that delatas are tried first and if
  they can't be applied it falls back to full state.
* For simultanious updates the messages may be reordered because we
  create separate WriteAggregator actor and such, but normally they
  will be sent in order so the deltas will typically be received in
  order, otherwise we fall back to retrying with full state in the
  second round in the WriteAggregator.

Refs #22188

### CODE REVIEW

**Comment by patriknw** [comment]
cc @gosubpl

**Comment by gosubpl** [comment]
Great!

File: akka-distributed-data/src/test/scala/akka/cluster/ddata/WriteAggregatorSpec.scala
Comment (johanandren) [inline akka-distributed-data/src/test/scala/akka/cluster/ddata/WriteAggregatorSpec.scala]: ok, so two writes to get to fullState2 gives two delta propagations here?

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: I think this was something rather unrelated that I found when testing something. `StoreFailure` should only be used for durable entries.

File: akka-distributed-data/src/test/scala/akka/cluster/ddata/WriteAggregatorSpec.scala
Comment (patriknw) [inline akka-distributed-data/src/test/scala/akka/cluster/ddata/WriteAggregatorSpec.scala]: I think this test only verifies that it tries with `DeltaPropagation` messages first.
`val nodes = Set(nodeA, nodeB, nodeC, nodeD)`
so it should send out a few such that are delivered to the `probe`

File: akka-distributed-data/src/main/protobuf/ReplicatorMessages.proto
Comment (patriknw) [inline akka-distributed-data/src/main/protobuf/ReplicatorMessages.proto]: fixed, thx

**Comment by patriknw** [comment]
I think @gosubpl has reviewed this so I take that as 2 LGTM. Will merge after another PR validation.

**Comment by gosubpl** [comment]
Yes, LGTM

**Pull request merged**
